### PR TITLE
[MPS] layernorm forward kernel

### PR DESCRIPTION
**Pull request by Isalia20**
Implements layernorm forward pass as a metal kernel instead of MPSGraph ops. Speed ups are indicated on the chart below:
Figure_1

Script for generating times, need to build torch with old/new codebase and then run this with different file name indicated at the end of the script
```python
import csv
import time

import numpy as np

import torch
import torch.nn.functional as F


matrix_sizes = [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192]
batch_sizes = [1]
elementwise_affine = [False, True]
num_runs = 50
warmup_runs = 3


def create_input_tensor(n, batch_size):
    torch.manual_seed(42)
    return torch.randn(batch_size, n, dtype=torch.float32)


def run_layer_norm(A, normalized_shape, elementwise_affine):
    torch.mps.synchronize()
    start = time.perf_counter()
    out = F.layer_norm(A, normalized_shape)
    torch.mps.synchronize()
    end = time.perf_counter()
    return out, end - start


results = {"N": [], "elementwise_affine": [], "batch_size": [], "mean_time": [], "std_time": []}

for el_aff in elementwise_affine:
    for n in matrix_sizes:
        for batch_size in batch_sizes:
            print(f"\nBenchmarking LayerNorm for input size N={n}, batch_size={batch_size}, elementwise_affine={el_aff}")

            try:
                A_cpu = create_input_tensor(n, batch_size)
                A_mps = A_cpu.to("mps")

                normalized_shape = (n,)

                for _ in range(warmup_runs):
                    _, _ = run_layer_norm(A_mps, normalized_shape, el_aff)

                times = []
                for _ in range(num_runs):
                    _, t = run_layer_norm(A_mps, normalized_shape, el_aff)
                    times.append(t)

                mean_time = np.mean(times)
                std_time = np.std(times)

                results["N"].append(n)
                results["elementwise_affine"].append(el_aff)
                results["batch_size"].append(batch_size)
                results["mean_time"].append(mean_time)
                results["std_time"].append(std_time)

                print(f"Mean time: {mean_time:.4f}s ± {std_time:.4f}s")

            except RuntimeError as e:
                print(f"Error for N={n}, batch_size={batch_size}: {e}")
                continue

with open("layernorm_benchmark_times_new.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerow(["N", "elementwise_affine", "batch_size", "mean_time", "std_time"])
    for i in range(len(results["N"])):
        writer.writerow(
            [
                results["N"][i],
                results["elementwise_affine"][i],
                results["batch_size"][i],
                results["mean_time"][i],
                results["std_time"][i],
            ]
        )

```

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 1]
```diff
@@ -878,6 +885,11 @@ static string get_mem_string(c10::MemoryFormat memory_format) {
                                                   const std::optional<Tensor>& weight_opt,
                                                   const std::optional<Tensor>& bias_opt,
                                                   double eps) {
+  int64_t N = std::accumulate(
+      normalized_shape.begin(), normalized_shape.end(), static_cast<int64_t>(1), std::multiplies<int64_t>());
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Normalization.mm:889]: Nit
```suggestion
  auto N = c10::multiply_integers(normalized_shape.begin(), normalized_shape.end());
```

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 2]
```diff
@@ -878,6 +885,11 @@ static string get_mem_string(c10::MemoryFormat memory_format) {
...
+  Tensor out = at::empty_like(input, MemoryFormat::Contiguous);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Normalization.mm:890]:
```suggestion
  auto out = at::empty_like(input, MemoryFormat::Contiguous);
```

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 3]
```diff
@@ -887,37 +899,58 @@ static string get_mem_string(c10::MemoryFormat memory_format) {
   auto M = M_N.first;
   auto X = input.expect_contiguous();
   auto gamma = weight.expect_contiguous();
+  Tensor mean = at::empty(batch_shape, input.options(), MemoryFormat::Contiguous);
+  Tensor rstd = at::empty(batch_shape, input.options(), MemoryFormat::Contiguous);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Normalization.mm:903]:
```suggestion
  auto mean = at::empty(batch_shape, input.options(), MemoryFormat::Contiguous);
  auto rstd = at::empty(batch_shape, input.options(), MemoryFormat::Contiguous);
```

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 4]
```diff
@@ -887,37 +899,58 @@ static string get_mem_string(c10::MemoryFormat memory_format) {
...
+      } else {
+        layerNormKernel = mps::lib.getPipelineStateForFunc("layer_norm_looped_" + metalType);
+      }
+      id<MTLComputeCommandEncoder> computeEncoder = stream->commandEncoder();
+      [computeEncoder setComputePipelineState:layerNormKernel];
+
+      [computeEncoder setBuffer:mps::getMTLBufferStorage(*X) offset:0 atIndex:0];
+      [computeEncoder setBuffer:mps::getMTLBufferStorage(out) offset:0 atIndex:1];
+      [computeEncoder setBuffer:mps::getMTLBufferStorage(mean) offset:0 atIndex:2];
+      [computeEncoder setBuffer:mps::getMTLBufferStorage(rstd) offset:0 atIndex:3];
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Normalization.mm:932]:
```suggestion
      mtl_setArgs(computeEncoder, X, out, mean, rstd);
```

**Review by malfet — APPROVED** [review]
